File: Backend/app/schemas/notices.py
"""
POST /api/notices/generate  요청/응답 스키마
POST /api/notices/send      요청/응답 스키마
POST /api/notices/refine    요청/응답 스키마
GET  /api/notices/{id}      응답 스키마
"""

from typing import Literal
from pydantic import BaseModel, Field


# ─── 요청 ─────────────────────────────────────────────────────────────────────

class StructuredStatus(BaseModel):
    health:     Literal["GOOD", "NORMAL", "NEEDS_OBSERVATION"]
    mood:       Literal["GOOD", "NORMAL", "ANXIOUS"]
    meal:       Literal["FULL", "NORMAL", "LITTLE", "REFUSED"]
    medication: Literal["DONE", "NONE"]


class GenerateRequest(BaseModel):
    residentId:             int
    participatedProgramIds: list[int] = Field(default_factory=list)
    status:                 StructuredStatus
    memo:                   str = ""
    tone:                   Literal["FRIENDLY", "POLITE", "EMPATHETIC"] = "POLITE"
    # 2차: 메모 입력 언어. None 이면 서버에서 자동 감지.
    memoLang:               str | None = None


# ─── 응답 ─────────────────────────────────────────────────────────────────────

class DraftOut(BaseModel):
    index: int          # 0 / 1 / 2
    label: str          # "A" | "B" | "C"
    text:  str


class AppliedProgram(BaseModel):
    programId: int
    title:     str
    startTime: str      # "HH:MM"


class GenerateResponse(BaseModel):
    drafts:          list[DraftOut]
    softenedCount:   int
    appliedPrograms: list[AppliedProgram]
    # 2차: 감지된 메모 입력 언어 (ko/vi/zh/en/기타)
    detectedLang:    str


# =============================================================================
# POST /api/notices/send
# =============================================================================

class SendRequest(BaseModel):
    residentId:             int
    rawMemo:                str = ""
    status:                 StructuredStatus
    participatedProgramIds: list[int] = Field(default_factory=list)
    tone:                   Literal["FRIENDLY", "POLITE", "EMPATHETIC"]
    aiGeneratedTexts:       list[DraftOut] = Field(min_length=1)   # 3안 전체 보존
    selectedDraftIndex:     int = Field(ge=0, le=2)                # 0/1/2
    isRefined:              bool
    finalText:              str = Field(min_length=1)              # 발송 본문 (필수)
    previousNoticeId:       int | None = None                       # 재전송이면 직전 notice id
    # 2차: 이 알림장 메모의 입력 언어 (notice.memo_lang 에 저장)
    memoLang:               str = "ko"


class NoticeSentOut(BaseModel):
    id:       int
    version:  int
    status:   Literal["SENT"]
    isEdited: bool
    sentAt:   str        # ISO 8601


class SendResponse(BaseModel):
    notice: NoticeSentOut


# =============================================================================
# POST /api/notices/refine
# =============================================================================

class RefineRequest(BaseModel):
    text: str = Field(min_length=1)
    tone: Literal["FRIENDLY", "POLITE", "EMPATHETIC"] = "POLITE"


class RefineResponse(BaseModel):
    refinedText: str
    changed:     bool


# =============================================================================
# GET /api/notices/{id}
# =============================================================================

class NoticeDetailOut(BaseModel):
    id:                   int
    residentId:           int
    version:              int
    status:               Literal["DRAFT", "SENT"]
    isEdited:             bool
    structuredStatus:     dict                    # JSONB 그대로
    participatedPrograms: list                    # JSONB 그대로 [{program_id,title,start_time}]
    rawMemo:              str | None
    tone:                 Literal["FRIENDLY", "POLITE", "EMPATHETIC"]
    selectedDraftIndex:   int | None
    isRefined:            bool
    finalText:            str | None
    sentAt:               str | None              # ISO 8601
    readAt:               str | None              # ISO 8601


class NoticeDetailResponse(BaseModel):
    notice: NoticeDetailOut

File: frontend/src/api/notices.ts
/**
 * 알림장 API
 * GET  /api/notices/:id
 * POST /api/notices/generate
 * POST /api/notices/refine
 * POST /api/notices/send
 */
import { api } from './client'
import {
  type Draft,
  type GenerateResponse,
  type Lang,
  type NoticeDetailResponse,
  type NoticeSentResponse,
  type RefineResponse,
  type StructuredStatus,
  type Tone,
} from '../types'

export function fetchNotice(id: number): Promise<NoticeDetailResponse> {
  return api.get<NoticeDetailResponse>(`/api/notices/${id}`)
}

export interface GeneratePayload {
  residentId: number
  participatedProgramIds: number[]
  status: StructuredStatus
  memo: string
  tone: Tone
  memoLang?: Lang        // 2차: 미전송 시 서버 자동 감지
}

export function generateDrafts(payload: GeneratePayload): Promise<GenerateResponse> {
  return api.post<GenerateResponse>('/api/notices/generate', payload)
}

export function refineText(text: string, tone: Tone): Promise<RefineResponse> {
  return api.post<RefineResponse>('/api/notices/refine', { text, tone })
}

export interface SendPayload {
  residentId: number
  rawMemo: string
  status: StructuredStatus
  participatedProgramIds: number[]
  tone: Tone
  aiGeneratedTexts: Draft[]
  selectedDraftIndex: number
  isRefined: boolean
  finalText: string
  previousNoticeId: number | null
  memoLang: Lang         // 2차: notice.memo_lang 저장용
}

export function sendNotice(payload: SendPayload): Promise<NoticeSentResponse> {
  return api.post<NoticeSentResponse>('/api/notices/send', payload)
}
